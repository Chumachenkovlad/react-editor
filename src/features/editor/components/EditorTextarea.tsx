import * as React from "react";
import { get } from "lodash";
import { EDITABLE_CLASS } from "../constants";

export interface IEditorTextareaProps {
  editor: {
    selectWord(el: HTMLElement): void;
  };
}

export default function EditorTextarea({
  editor
}: IEditorTextareaProps): React.ReactElement {
  function selectWord() {
    const selection = window.getSelection();

    if (selection === null) {
      return;
    }

    const parent = get(selection, "anchorNode.parentElement") as HTMLElement;

    if (parent.classList.contains(EDITABLE_CLASS)) {
      editor.selectWord(parent);
    } else {
      const range = selection.getRangeAt(0);
      const container = document.createElement("span");
      container.classList.add(EDITABLE_CLASS);
      range.surroundContents(container);
      editor.selectWord(container);
    }
  }

  return (
    <div contentEditable onDoubleClick={selectWord}>
      Subjectivity is very important when considering privacy because we are
      seeing diverse levels of tolerance towards what is considered acceptable
      and what are privacy infringements. Some may claim that privacy is dead,
      or as Mark Zuckerburg said in 2010 that ‘Privacy was no longer a social
      norm.’ Although to be fair Mark Zuckerberg he did also say in a Times
      interview also in 2010 that “What people want isn’t complete privacy. It
      isn’t that they want secrecy. It’s that they want control over what they
      share and what they don’t.
    </div>
  );
}
