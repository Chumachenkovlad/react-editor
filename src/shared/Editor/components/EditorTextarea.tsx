import * as React from "react";
import { get, debounce } from "lodash";
import { EDITABLE_CLASS } from "../constants/constants";
import { findOuterEditableNode } from '../helpers/find-outer-editable-node.func';

export interface IEditorTextareaProps {
  selectWord(el: HTMLElement): void;
  children: JSX.Element;
}

const SELECTION_DEBOUNCE = 150;

export default function EditorTextarea(
  props: IEditorTextareaProps
): React.ReactElement {
  function selectWord() {
    const selection = window.getSelection();

    if (!selection) {
      return;
    }

    const parent = findOuterEditableNode(
      get(selection, "anchorNode.parentNode") as HTMLElement
    );

    if (!parent) {
      return;
    }

    if (parent.classList.contains(EDITABLE_CLASS)) {
      return props.selectWord(parent);
    }

    const range = selection.getRangeAt(0);
    const container = document.createElement("span");
    container.classList.add(EDITABLE_CLASS);
    range.surroundContents(container);

    if (container.innerText) {
      return props.selectWord(container);
    }

    container.remove();
  };

  const debouncedSelectWord = debounce(selectWord, SELECTION_DEBOUNCE)

  return (
    <div
      className="EditorTextarea"
      suppressContentEditableWarning={true}
      contentEditable
      onMouseUp={debouncedSelectWord}
    >
      {props.children}
    </div>
  );
}
