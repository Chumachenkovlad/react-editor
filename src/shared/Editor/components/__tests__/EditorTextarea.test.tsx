import { render, RenderResult } from '@testing-library/react';
import EditorTextarea from '../EditorTextarea';
import React from "react";

const TEXT = "some extra text";
describe('<EditorTextarea>', () => {
    let EditorTextareaRR: RenderResult;
    beforeEach(() => {
        EditorTextareaRR = render(<EditorTextarea selectWord={() => { }} children={<div>{TEXT}</div>} />)
    });

    test('should be defined', () => {
        expect(EditorTextareaRR.container).toMatchSnapshot();
    });

    test('should includes defined text', () => {
        expect(EditorTextareaRR.getByText(TEXT)).toBeInTheDocument();
    });
})
