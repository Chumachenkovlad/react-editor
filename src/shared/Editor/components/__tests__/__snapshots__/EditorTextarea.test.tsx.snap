// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditorTextarea> should be defined 1`] = `
<div>
  <div
    class="EditorTextarea"
    contenteditable="true"
  >
    <div>
      some extra text
    </div>
  </div>
</div>
`;
